sequencer/sequencer.go

```diff
@@ -0,0 +1,148 @@
+// Package sequencer provides functionality for processing and aggregating ballots
+// into batches with zero-knowledge proofs for efficient verification.
+package sequencer
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/constraint"
+	"github.com/vocdoni/vocdoni-z-sandbox/circuits/aggregator"
+	ballottest "github.com/vocdoni/vocdoni-z-sandbox/circuits/test/ballotproof"
+	"github.com/vocdoni/vocdoni-z-sandbox/circuits/voteverifier"
+	"github.com/vocdoni/vocdoni-z-sandbox/log"
+	"github.com/vocdoni/vocdoni-z-sandbox/storage"
+)
+
+// Sequencer is a worker that takes verified ballots and aggregates them into a single proof.
+// It processes ballots and creates batches of proofs for efficient verification.
+type Sequencer struct {
+	stg      *storage.Storage
+	ctx      context.Context
+	cancel   context.CancelFunc
+	pids     map[string]time.Time // Maps process IDs to their last update time
+	pidsLock sync.RWMutex         // Protects access to the pids map
+
+	ballotVerifyingKeyCircomJSON []byte // Verification key for ballot proofs
+
+	aggregateProvingKey groth16.ProvingKey          // Key for generating aggregate proofs
+	aggregateCcs        constraint.ConstraintSystem // Constraint system for aggregate proofs
+
+	voteProvingKey groth16.ProvingKey          // Key for generating vote proofs
+	voteCcs        constraint.ConstraintSystem // Constraint system for vote proofs
+
+	// maxTimeWindow is the maximum time window to wait for a batch to be processed.
+	// If this time elapses, the batch will be processed even if not full.
+	maxTimeWindow time.Duration
+}
+
+// New creates a new Sequencer instance that processes ballots and aggregates them into batches.
+// It loads all necessary cryptographic artifacts for proof verification and generation.
+//
+// Parameters:
+//   - stg: Storage instance for accessing ballots and other data
+//   - batchTimeWindow: Maximum time to wait before processing a batch even if not full
+//
+// Returns a configured Sequencer instance or an error if initialization fails.
+func New(stg *storage.Storage, batchTimeWindow time.Duration) (*Sequencer, error) {
+	if stg == nil {
+		return nil, fmt.Errorf("storage cannot be nil")
+	}
+	if batchTimeWindow <= 0 {
+		return nil, fmt.Errorf("batch time window must be positive")
+	}
+
+	// Load vote verifier artifacts
+	vvArtifacts := voteverifier.Artifacts
+	if err := vvArtifacts.LoadAll(); err != nil {
+		return nil, fmt.Errorf("failed to load vote verifier artifacts: %w", err)
+	}
+
+	// Decode the vote verifier circuit definition
+	voteCcs := groth16.NewCS(ecc.BLS12_377)
```

Comment:
you can replace those constants by: https://github.com/vocdoni/davinci-node/blob/main/circuits/curves.go